src/tokens/selectors.ts

```diff
@@ -42,21 +43,108 @@ const DEFAULT_MEMOIZE_MAX_SIZE = 10
 
 function isNetworkIdList(networkIds: any): networkIds is NetworkId[] {
   return (
-    networkIds.constructor === Array &&
+    Array.isArray(networkIds) &&
     networkIds.every((networkId) => Object.values(NetworkId).includes(networkId))
   )
 }
 export const tokenFetchLoadingSelector = (state: RootState) => state.tokens.loading
 export const tokenFetchErrorSelector = (state: RootState) => state.tokens.error
 
+// Note: not importing from 'src/positions/selectors' to avoid circular dependency
+// TODO: address circular dependency
+const positionsSelector = (state: RootState) => state.positions.positions
+const positionsFetchedAtSelector = (state: RootState) => state.positions.positionsFetchedAt
+
+const positionTokensSelector = createSelector([positionsSelector], (positions) => {
+  const positionTokens: Record<string, Token> = {}
+
+  function visitToken(token: Token, isTopLevelAppToken: boolean) {
+    const tokenId = token.tokenId
+    if (!positionTokens[tokenId] || isTopLevelAppToken) {
+      positionTokens[tokenId] = {
+        ...token,
+        // Only keep the balance if it's a top level app-token
+        // Otherwise it doesn't represent a real user balance
+        balance: isTopLevelAppToken ? token.balance : '0',
+      }
+    }
+    if (token.type === 'app-token') {
+      for (const childToken of token.tokens) {
+        visitToken(childToken, false)
+      }
+    }
+  }
+
+  for (const position of positions) {
+    if (position.type === 'app-token') {
+      visitToken(position, true)
+    } else {
+      for (const token of position.tokens) {
+        visitToken(token, false)
+      }
+    }
+  }
+  return positionTokens
+})
+
+type TokensByIdArgs =
+  | NetworkId[] // For backward compatibility
```

Comment:
we can eventually drop this as an option and we should be able to catch everything with TS errors right?

Reply:
yes exactly.